lease/lessor_bench_test.go

```diff
@@ -15,112 +15,207 @@
-	lg := zap.NewNop()
-	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
+func benchmarkLessorGrant(benchSize int, b *testing.B) {
+	ttls := randomTTL(benchSize, minTTL, maxTTL)
+
+	var le *lessor
+	var tearDown func()
+
+	b.ResetTimer()
+	for i := 0; i < b.N; {
+		b.StopTimer()
+		if tearDown != nil {
+			tearDown()
+			tearDown = nil
+		}
+		le, tearDown = setUp()
+		b.StartTimer()
+
+		for j := 1; j <= benchSize; j++ {
+			le.Grant(LeaseID(j), ttls[j-1])
+		}
+		i += benchSize
 	}
-	for i := 0; i < b.N; i++ {
-		le.Grant(LeaseID(i+size), int64(100+i+size))
+	b.StopTimer()
+
+	if tearDown != nil {
+		tearDown()
 	}
+}
+
+func benchmarkLessorRevoke(benchSize int, b *testing.B) {
+	ttls := randomTTL(benchSize, minTTL, maxTTL)
+
+	var le *lessor
+	var tearDown func()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		le.Revoke(LeaseID(i + size))
+		b.StopTimer()
+		if tearDown != nil {
+			tearDown()
+			tearDown = nil
+		}
+		le, tearDown = setUp()
+		for j := 1; j <= benchSize; j++ {
+			le.Grant(LeaseID(j), ttls[j-1])
+		}
+		b.StartTimer()
+
+		for j := 1; j <= benchSize; j++ {
+			le.Revoke(LeaseID(j))
+		}
+		i += benchSize
 	}
-}
+	b.StopTimer()
 
-func benchmarkLessorRenew(size int, b *testing.B) {
-	lg := zap.NewNop()
-	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
+	if tearDown != nil {
+		tearDown()
 	}
+}
+
+func benchmarkLessorRenew(benchSize int, b *testing.B) {
+	ttls := randomTTL(benchSize, minTTL, maxTTL)
+
+	var le *lessor
+	var tearDown func()
+
 	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		le.Renew(LeaseID(i))
+	for i := 0; i < b.N; {
+		b.StopTimer()
+		if tearDown != nil {
+			tearDown()
+			tearDown = nil
+		}
+		le, tearDown = setUp()
+		for j := 1; j <= benchSize; j++ {
+			le.Grant(LeaseID(j), ttls[j-1])
+		}
+		b.StartTimer()
+
+		for j := 1; j <= benchSize; j++ {
+			le.Renew(LeaseID(j))
+		}
+		i += benchSize
+	}
+	b.StopTimer()
+
+	if tearDown != nil {
+		tearDown()
 	}
 }
 
-func cleanup(b backend.Backend, path string) {
-	b.Close()
-	os.Remove(path)
+func benchmarkLessorFindExpired(benchSize int, b *testing.B) {
+	// 50% lease are expired.
+	ttls := randomTTL(benchSize, -500, 500)
+	findExpiredLimit := 50
+
+	var le *lessor
+	var tearDown func()
+
+	b.ResetTimer()
+	for i := 0; i < b.N; {
+		b.StopTimer()
+		if tearDown != nil {
+			tearDown()
+			tearDown = nil
+		}
+		le, tearDown = setUp()
+		for j := 1; j <= benchSize; j++ {
+			le.Grant(LeaseID(j), ttls[j-1])
+		}
+		// lessor's runLoop should not call findExpired
+		demote(le)
+		b.StartTimer()
+
+		// refresh fixture after pop all expired lease
+		for ;;i++ {
```

Comment:
Test failed due to formatting:
https://travis-ci.com/etcd-io/etcd/jobs/205296470#L588